File: stereo_cam/scripts/stereo_cam_publisher.cpp
#include "ros/ros.h"
#include "sensor_msgs/Image.h"
#include <cv_bridge/cv_bridge.h>
#include <image_transport/image_transport.h>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;


int main(int argc, char** argv)
{   
    //node init
    ros::init(argc, argv, "publish_camera");
    ros::NodeHandle nh;
    image_transport::ImageTransport it(nh);
    int img_w{1280};
    int img_h{480};
    int fps{30};
    int queue_size{10};
    bool auto_exposure{true};
    int exposure_time{-6};
    bool auto_white_balance{true};
    string pubstr{"/stereo_camera/raw"};
    string image_format{"mjpg"};
    cv_bridge::CvImage img_bridge;
    int img_counter{0};
    cv::Mat cv_frame;
    sensor_msgs::Image img_msg;
    string cam_port_name{"/dev/video0"};


    //load param
    nh.getParam("img_width", img_w);
    nh.getParam("img_height", img_h);
    nh.getParam("frame_rate", fps);
    nh.getParam("auto_exposure", auto_exposure);
    nh.getParam("exposure_time", exposure_time);
    nh.getParam("auto_white_balance", auto_white_balance);
    if(nh.hasParam("topic_name"))
    {
        nh.getParam("topic_name",pubstr);
    }
    if(nh.hasParam("image_format"))
    {
        nh.getParam("image_format",image_format);
    }
    if(nh.hasParam("cam_port_name"))
    {
        nh.getParam("cam_port_name",cam_port_name);
    }
    if(nh.hasParam("queue_size"))
    {
        nh.getParam("queue_size", queue_size); 
    }


    //camera init
    //cv::VideoCapture cap;
    //cap.open(0);
    //cap.open(0);
    cv::VideoCapture cap(cam_port_name,cv::CAP_V4L2);
    if (image_format == "mjpg")
        {std::cout << "Image Format :MJPG: " << cap.set(CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));}
    else
        {std::cout << "Image Format :YUYV: " << cap.set(CAP_PROP_FOURCC, cv::VideoWriter::fourcc('Y', 'U', 'Y', 'V'));}
    //cap.set(CAP_PROP_FPS,fps);
    cap.set(3,img_w);
    cap.set(4, img_h);
    std::cout<<"fps: "<< cap.get(5);
    //std::cout<<"format: "<< cap.get(CAP_PROP_FOURCC);
    std::cout<<"img_w: "<< cap.get(3);
    std::cout<<"img_h: "<< cap.get(4);
    if(!cap.isOpened())
        {
            ROS_ERROR("Camera is not opened\n");
        }
    else
    {
        ROS_INFO("Camera opened\n");
    }
    cap.set(CAP_PROP_AUTO_EXPOSURE,3);
    if(!auto_exposure)
    {
        std::cout<<"EXPOSURE RESET"<<endl;
        //cap.set(CAP_PROP_AUTO_EXPOSURE,3);
        cap.set(CAP_PROP_AUTO_EXPOSURE,1);
        cap.set(CAP_PROP_EXPOSURE,exposure_time);
    }
    cap.set(CAP_PROP_AUTO_WB,3);
    if(!auto_white_balance)
    {
        std::cout<<"WHITEBALANCE RESET"<<endl;
        cap.set(CAP_PROP_AUTO_WB,1);
        cap.set(CAP_PROP_WB_TEMPERATURE,6500);
    }
    std::cout<<"Auto_Exposure"<<cap.get(CAP_PROP_AUTO_EXPOSURE)<<endl;
    std::cout<<"Exposure_Time"<<cap.get(CAP_PROP_EXPOSURE)<<endl;
    std::cout<<"AUTO_WHITEBALANCE"<<cap.get(CAP_PROP_AUTO_WB)<<endl;
    std::cout<<"WB_TEMP"<<cap.get(CAP_PROP_WB_TEMPERATURE)<<endl;
    //publisher init
    image_transport::Publisher rawpub = it.advertise(pubstr,queue_size);


    //start loop
    ros::Rate loop_rate(fps);
    while (ros::ok())
        {   
            if(!cap.read(cv_frame))
            {
                ROS_WARN("Detect no Image\n");
            }
            else
            {   //cap >> cv_frame;
                //std::cout<<"size"<<cv_frame.size()<<endl;
                cv::cvtColor(cv_frame, cv_frame, CV_BGR2RGB);
                std_msgs::Header header; // empty header
                header.seq = img_counter++; // user defined counter
                header.stamp = ros::Time::now(); // time
                //std::cout<<header.stamp<<endl;
                img_bridge = cv_bridge::CvImage(header, sensor_msgs::image_encodings::RGB8, cv_frame);
                img_bridge.toImageMsg(img_msg);
                rawpub.publish(img_msg);

            }

            ros::spinOnce();
            loop_rate.sleep();
        }

    return 0;
}

File: stereo_cam/scripts/stereo_cam_publisher.py
#import numpy as np
import cv2
import os
import rospy
from cv_bridge import CvBridge
#https://stackoverflow.com/questions/73400055/ros-melodic-image-from-webcam


from std_msgs.msg import String
from sensor_msgs.msg import Image


class Camera(object):

    def __init__(self):
        #self.height = 1280
        #self.width = 480
        #self.fps = 10

        self.bridge = CvBridge()
        self.img_w = rospy.get_param('img_width', default=1280)
        self.img_h = rospy.get_param('img_height', default=480)
        self.fps = rospy.get_param('frame_rate', default=30)
        self.auto_exp = rospy.get_param('auto_exposure', default=True)
        #https://blog.csdn.net/smartvxworks/article/details/108792343
        self.exp_level = rospy.get_param('exposure_time', default=-6) 
        self.auto_wb = rospy.get_param('auto_white_balance', default=True)
        

        self.pubstr = "/stereo_camera/raw"
        self.rawpub = rospy.Publisher(self.pubstr, Image, queue_size=10)
        self.cap = cv2.VideoCapture(-1)
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH,self.img_w)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT,self.img_h)
        self.cap.set(cv2.CAP_PROP_FPS, self.fps)
        if not self.auto_exp:
            self.cap.set(cv2.CAP_PROP_AUTO_EXPOSURE, 0.25) # manual mode
            self.cap.set(cv2.CAP_PROP_EXPOSURE, self.exp_level)
        self.cap.set(cv2.CAP_PROP_AUTO_WB, self.auto_wb) 
        
        output_info_exp = "For detail of the exposure level, please refer to {}".format('https://blog.csdn.net/smartvxworks/article/details/108792343')
        output_info_cam = "img_w={}\timg_h={}\timg_fps={}\tauto_exposure={}\texposure_level={}\tauto_white_balance={}".format(
                        self.img_w,self.img_h,self.fps,
                        self.auto_exp, self.exp_level, self.auto_wb)
        rospy.loginfo('\n\n'+'*'*15+'CAMERA INFO'+'*'*15 + '\n' + output_info_exp + '\n' + output_info_cam + '\n' + '*'*15+'CAMERA INFO'+'*'*15+'\n\n')


    def run(self):
        if not self.cap.isOpened():
            print("Camera is not opened\n")

        ret, cv_image = self.cap.read()
        #print(cv_image.shape)
        #left_frame = cv_frame[0:480, 0:640]
        #right_frame = cv_frame[0:480, 640:1280]

        if not ret:
            print("Detect no Image\n")

        else:
            #export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtiff.so.5
            self.rawpub.publish(self.bridge.cv2_to_imgmsg(cv_image, encoding="bgr8"))

if __name__ == '__main__':
    rospy.init_node("Publish_camera")
    camera = Camera()
    r = rospy.Rate(30)

    while not rospy.is_shutdown():
        camera.run()
        r.sleep()